### ISSUE TITLE
REAL16 proposal

### ISSUE THREAD

**Issue by certik** [lfortran/fortran_proposals#13]
The document [N2169](https://isotc.iso.org/livelink/livelink?func=ll&objId=20646090&objAction=Open) lists the approved items from the WG5 August 2019 Meeting in Tokyo, one of which is:

US08 - Additional named constant, REAL16, in ISO_FORTRAN_ENV to specify whether the processor supports a 16-bit REAL KIND. [J3/19-139r1](https://j3-fortran.org/doc/year/19/19-139r1.txt)

**Comment by klausler** [comment]
Why not use `SELECTED_REAL_KIND()`?

```
SELECTED_REAL_KIND(3, 4) ! is IEEE 16-bit
SELECTED_REAL_KIND(2, 37) ! is truncated IEEE 32-bit
SELECTED_REAL_KIND(6, 37) ! is IEEE 32-bit
SELECTED_REAL_KIND(15, 307) ! is IEEE 64-bit
SELECTED_REAL_KIND(18, 4931) ! is x87 80-bit
SELECTED_REAL_KIND(32, 307) ! is "double-double"
SELECTED_REAL_KIND(33, 9863) ! is IEEE 128-bit
```

These are encoded with maximum values, not minima, and perhaps they'd be more clear if written so.  Exercise for the reader.

Anyway, if `SELECTED_REAL_KIND(2,4) == SELECTED_REAL_KIND(6, 37)`, the compiler (ahem processor) does not have a 16-bit `REAL`.

**Comment by certik** [comment]
I know there was a discussion about this point, as I asked the same question. @tclune do you remember what the arguments were?

**Comment by tclune** [comment]
I think I've missed something.   What is the question/position for which arguments are desired?

**Comment by certik** [comment]
@klausler was asking why not to just use `SELECTED_REAL_KIND` instead of `REAL16` from `ISO_FORTRAN_ENV`.

**Comment by tclune** [comment]
Sorry  - I did not real the OP carefully enough.  Now I understand the context.

At the very least checking to see if two results from SELECTED_REAL_KIND differ is something that happens at run time, and sometimes one wants to know at compile time that some functionality is not available. 

The question is really about why the existing standard has REAL32, REAL64, and REAL128.   These are much more practical for library developers that wish to overload their procedures for the supported kinds of REAL provided by a given implementation.       In that case it is not that the developer wants a certain amount of precision.   They are trying to support other users that get to choose precision on their own.     REAL16 is proposed just simply to be consistent for an increasingly common option on some hardware.

Note that REAL32 et al are not quite sufficient in and of themselves to allow a portable implementation that covers all supported real kinds.    Generally some logic is still required in the build system to control which kinds are selected.

Note also that REAL32 is not necessarily the same thing as default REAL.  Theoretically it can be tricky to distinguish such things.  In practice, all extant compilers have ```kind(1.) == REAL32```

**Comment by klausler** [comment]
> Sorry - I did not real the OP carefully enough. Now I understand the context.
> 
> At the very least checking to see if two results from SELECTED_REAL_KIND differ is something that happens at run time, and sometimes one wants to know at compile time that some functionality is not available.

`SELECTED_REAL_KIND()` can appear in a constant expression when its arguments are constant expressions (10.1.12(1)(6)).

**Comment by FortranFan** [comment]
@klausler wrote:

> Why not use `SELECTED_REAL_KIND()`?
> ..

@klausler, can you please explain your point here?

Are you giving a "thumbs down" or suggesting the proposal to introduce the proposal involving a trivial effort to add another few named constants (e.g., REAL16, LOGICAL8, LOGICAL16, ..)) to the list of those introduced in the language starting Fortran 2008 (such as REAL32) be removed from consideration?

**Comment by certik** [comment]
@FortranFan I think @klausler is just asking a general question if user code should be using `SELECTED_REAL_KIND` or `ISO_FORTRAN_ENV`. Thanks @tclune for the answer.

**Comment by klausler** [comment]
`SELECTED_REAL_KIND()` works better, IMO.  It won't be ambiguous like `REAL128` may become if Power9 supports both IEEE 128-bit and "double-double", or like `REAL16` will be on systems supporting both standard  IEEE 16-bit and the truncated form of 32-bit ("bfloat16").  I'm not saying that Fortran should remove these (and the others) from ISO_FORTRAN_ENV, but I discourage their use.

**Comment by certik** [comment]
@klausler this is a long standing issue in Fortran, how you are supposed to declare the precision of the `real` variable. With multiple opinions even on the committee itself. I personally actually just use `dp = kind(1.d0)` for double precision, as that is the shortest, and otherwise I use  `SELECTED_REAL_KIND`. I do not use `ISO_FORTRAN_ENV`. But I know several colleagues that use `ISO_FORTRAN_ENV`.

**Comment by klausler** [comment]
`SELECTED_REAL_KIND()` is pretty ingenious -- a compile-time constant function for finding the best fitting real type in the target environment based on actual requirements on precision and range.  Standard C/C++ don't have it.  Code porting will always require tuning for performance and numerical stability, but at least with `SELECTED_REAL_KIND()` you don't have to know quite as much about the target architecture.

**Comment by certik** [comment]
@klausler exactly. The other functions that go hand in hand with this are `epsilon` and `tiny`, both of which I use quite often. That way the numerical code is actually quite independent of the actual accuracy of the real type. In practice I only test with double precision, but I think the porting to single precision or quadruple precision would not be that hard with my codes.
